openedx_external_enrollments/external_enrollments/pathstream_external_enrollment.py

```diff
@@ -0,0 +1,201 @@
+from django.conf import settings
+
+from openedx_external_enrollments.external_enrollments.base_external_enrollment import BaseExternalEnrollment
+from openedx_external_enrollments.models import ExternalEnrollment
+
+LOG = logging.getLogger(__name__)
+COMPLETED = True
+UNCOMPLETED = False
+
+
+class S3NotInitialized(BaseException):
+    """Exception when _init_s3 has not been initialized before calling _upload_file or _download_file """
+
+
+class PathstreamExternalEnrollment(BaseExternalEnrollment):
+    """
+    PathstreamExternalEnrollment class.
+    """
+    def __init__(self):
+        self.client = None
+        self.S3_BUCKET = settings.OEE_PATHSTREAM_S3_BUCKET
+        self.S3_FILE = settings.OEE_PATHSTREAM_S3_FILE
+
+    def __str__(self):
+        return 'pathstream'
+
+    def _init_s3(self):
+        """Start S3 connection.
+        """
+        self.client = boto3.client(
+            's3',
+            aws_access_key_id=settings.OOE_PATHSTREAM_S3_ACCESS_KEY,
+            aws_secret_access_key=settings.OOE_PATHSTREAM_S3_SECRET_KEY,
+        )
+
+    def _get_enrollment_data(self, data, course_settings):
+        """
+        Returns a string with the data required to be treated as a log.
+        String format: 'course_key,email,date_time,status'
+        """
+        date_time = datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S.%f")
+
+        enrollment_data = u'{course_key},{email},{date_time},{status}\n'.format(
+            course_key=data.get('course_id'),
+            email=data.get('user_email'),
+            date_time=date_time,
+            status=str(data.get('is_active')).lower(),
+        )
+        return enrollment_data
+
+    def _post_enrollment(self, data, course_settings=None):
+        """
+        Save enrollment data in EnrollmentRequestLog.
+        """
+        enrollment_data = self._get_enrollment_data(data, course_settings)
+        LOG.info('calling enrollment for [%s] with data: %s', self.__str__(), data)
+        LOG.info('calling enrollment for [%s] with course settings: %s', self.__str__(), course_settings)
+
+        details = {
+            'enrollment_data_formated': enrollment_data,
+            'course_advanced_settings': course_settings,
+            'is_uploaded': False,
+        }
+
+        enrollment = ExternalEnrollment.objects.create(  # pylint: disable=no-member
+            controller_name=str(self),
+            course_shell_id=data.get('course_id'),
+            email=data.get('user_email'),
+            meta=details
+        )
+
+        LOG.info('Saving External enrollment object for [%s]'
+                 ' -- ExternalEnrollment.id = %s', self.__str__(), enrollment.id)
+
+    def _download_file(self):
+        """Download the file from an S3 Bucket.
+
+        :param file_name: File to download
+        :param bucket: Bucket to download from
+
+        :return: True if file was downloaded, else False"""
+        if self.client is None:
+            raise S3NotInitialized
```

Comment:
I believe we can handle this exception in the next try except block, what do you think @Jacatove ?

Reply:
Let's suppose that we handle the exception in the next try except block as you suggest. In case `self.client = None` it will throw an AttributeError because `None` does not have the `download_file`, so the code would looks like this next block of code, that correct?
```
        try:
            self.client.download_file(
                self.S3_BUCKET,
                self.S3_FILE,
                self.S3_FILE
            )
        except ClientError as error:
            LOG.error('Failed to download the file to S3. Reason: %s', str(error))
            return False
        except AttributeError:
            raise S3NotInitialized
```
But I think that future readers of the code might find the reason of why is mandatory to call `_init_s3` before, If we keep it at the beginning of the function.